src/backend/native_gpu_future.rs

```diff
@@ -24,43 +28,58 @@ impl<T> Future for GpuFuture<T>
     type Output = T;
 
     fn poll(self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
-        // grab a clone of the Arc
-        let arc = Arc::clone(&self.get_mut().inner);
+        let inner = &self.get_mut().inner;
 
         // grab the device id and set the waker, but release the lock, so that the native callback can write to it
         let device_id = {
-            let mut inner = arc.lock().unwrap();
-            inner.waker.replace(context.waker().clone());
+            let mut inner = inner.lock().unwrap();
+
+            if let Some(WakerOrResult::Result(res)) = inner.waker_or_result.take() {
+                return Poll::Ready(res);
+            }
+
+            inner.waker_or_result.replace(WakerOrResult::Waker(context.waker().clone()));
+
             inner.id
         };
 
         // polling the device should trigger the callback
-        wgn::wgpu_device_poll(device_id, true);
+        wgn::wgpu_device_poll(device_id, false);
 
         // now take the lock again, and check whether the future is complete
-        let mut inner = arc.lock().unwrap();
-        match inner.result.take() {
-            Some(value) => Poll::Ready(value),
-            _ => Poll::Pending,
+        let mut inner = inner.lock().unwrap();
+
+        match inner.waker_or_result.take() {
+            Some(WakerOrResult::Result(res)) => Poll::Ready(res),
+            value @ _ => {
+                inner.waker_or_result = value;
+                Poll::Pending
+            },
         }
     }
 }
 
 impl<T> GpuFutureCompletion<T> {
     pub fn complete(self, value: T) {
         let mut inner = self.inner.lock().unwrap();
-        inner.result.replace(value);
-        if let Some(waker) = &inner.waker {
-            waker.wake_by_ref();
-        }
+
+        match inner.waker_or_result.take() {
+            Some(WakerOrResult::Waker(waker)) => {
+                inner.waker_or_result.replace(WakerOrResult::Result(value));
+                waker.wake();
+            },
+            None => {
+                inner.waker_or_result.replace(WakerOrResult::Result(value));
+            }
+            Some(WakerOrResult::Result(_)) => unreachable!(),
```

Comment:
is it truly unreachable? or just unexpected?

Reply:
I think it's truly unreachable. The `Result` case will only be set once `complete` is called, and since it can only be called once because it's consuming, that case will never happen.

Reply:
but this is a public function of a public type, no?

Reply:
Yes, but the invariants are checked here. There's no way that `waker_or_result` can be `Some(WakerOrResult::Result(...))` before `complete` is called, so that case will never be reached.